source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -1999,59 +2000,139 @@ void COpenGLDriver::SAuxContext::flushState(GL_STATE_BITS stateBits)
     }
 #undef STATE_NEQ
 #undef UPDATE_STATE
-    if ((stateBits & GSB_DESCRIPTOR_SETS) && currentState.pipeline)
+    if (stateBits & GSB_DESCRIPTOR_SETS)
     {
-        //TODO get rid of constants and use runtime-queried binding point counts
-        GLuint ubonames[MAX_UBO_BINDING_COUNT]{};
-        GLintptr ubooffsets[MAX_UBO_BINDING_COUNT]{};
-        GLsizeiptr ubosizes[MAX_UBO_BINDING_COUNT]{};
-
-        GLuint ssbonames[MAX_SSBO_BINDING_COUNT]{};
-        GLintptr ssbooffsets[MAX_SSBO_BINDING_COUNT]{};
-        GLsizeiptr ssbosizes[MAX_SSBO_BINDING_COUNT]{};
-
-        GLuint texnames[MAX_TEXTURE_BINDING_COUNT]{};
-        GLuint smplrnames[MAX_TEXTURE_BINDING_COUNT]{};
-
-        GLuint imgnames[MAX_IMAGE_BINDING_COUNT]{};
-        GLenum imgformats[MAX_IMAGE_BINDING_COUNT]{};
+        //provide storage for multibind calls' parameters
+        constexpr size_t UNIT_SIZE = 5*sizeof(GLuint) + 2*sizeof(GLintptr) + 2*sizeof(GLsizeiptr);
+        uint8_t stackmem[1u<<14]{};//16k bytes
+        constexpr size_t ARRAY_SIZE = sizeof(stackmem)/UNIT_SIZE;
+
+#define ASSIGN_MEM(arrayname) reinterpret_cast<decltype(arrayname)>(stackmem+offset)
+#define UPDATE_OFFSET(arrayname) offset += ARRAY_SIZE*sizeof(*arrayname)
+        size_t offset = 0ull;
+        GLuint* const ubonames = ASSIGN_MEM(ubonames);
+        UPDATE_OFFSET(ubonames);
+        GLintptr* const ubooffsets = ASSIGN_MEM(ubooffsets);
+        UPDATE_OFFSET(ubooffsets);
+        GLsizeiptr* const ubosizes = ASSIGN_MEM(ubosizes);
+        UPDATE_OFFSET(ubosizes);
+
+        GLuint* const ssbonames = ASSIGN_MEM(ssbonames);
+        UPDATE_OFFSET(ssbonames);
+        GLintptr* const ssbooffsets = ASSIGN_MEM(ssbooffsets);
+        UPDATE_OFFSET(ssbooffsets);
+        GLsizeiptr* const ssbosizes = ASSIGN_MEM(ssbosizes);
+        UPDATE_OFFSET(ssbosizes);
+
+        GLuint* const texnames = ASSIGN_MEM(texnames);
+        UPDATE_OFFSET(texnames);
+        GLuint* const smplrnames = ASSIGN_MEM(smplrnames);
+        UPDATE_OFFSET(smplrnames);
+
+        GLuint* const imgnames = ASSIGN_MEM(imgnames);
+#undef UPDATE_OFFSET
+#undef ASSIGN_MEM
+
+        //unbind previous descriptor sets
```

Comment:
you can do bind new and unbind old at the same time.

Reply:
see https://github.com/buildaworldnet/IrrlichtBAW/pull/358#discussion_r333204559